### ISSUE TITLE
[RNN] TFLite integer 8 quantization yielding worse accuracy than CoreML integer 8

### ISSUE THREAD

**Issue by lucacampanella** [tensorflow/tensorflow#48487]
### 1. System information

- OS Platform and Distribution: Linux Ubuntu 20.04
- TensorFlow installation: pip package
- TensorFlow library: tf-nightly==2.6.0.dev20210407

### 2. Issue and code

I'm implementing a Keras GRU model (tensorflow.keras.layers.GRU) on mobile devices for character level language modelling. In order to have a smaller on-device checkpoint I'm experimenting both with TFLite and CoreML integer quantization at 8 bits. The problem is that the TFLite compressed model yields up to 3% worse accuracy results on the test set when compared to the CoreML compressed model. The test set is big enough and the CoreML model results are very close to the full 32 bits model (or the 16 bit model).

This is how I'm executing the conversion for TFLite (where model is of type `tf.keras.Model`):

```
converter = tf.lite.TFLiteConverter.from_keras_model(model)

converter.experimental_new_converter = True
converter.experimental_new_quantizer = True
converter.optimizations = [tf.lite.Optimize.DEFAULT]

converted = converter.convert()
```

I've tried with various versions of TensorFlow, including the 2.4.1, 2.5.0 and the nightly.

This is how I'm executing the conversion for CoreML (where model is of type `tf.keras.Model`):

```
import coremltools as ct

coreml_model = ct.convert(model, use_float_arraytype=True)
coreml_model = ct.models.neural_network.quantization_utils.quantize_weights(coreml_model, 8, 
                                                                     quantization_mode="linear")
```
TensorFlow Lite doesn't allow you to specify a quantization mode, while CoreML lets you choose among: linear (default), linear_lut, kmeans_lut, custom_lut, linear_symmetric. I tried all of them but custom_lut. The ones I've tried all yield results very similar to the un-quantized model, except for linear_lut.

Now the question is: why is the model coming out of the TFLite quantization process so much worse than almost all the approaches coming from CoreML? is there a way to reproduce the CoreML results in TFLite?

Thanks a lot!

**Comment by abattery** [comment]
@Xhark could you triage this?

**Comment by abattery** [comment]
@teijeong could you triage this issue?

**Comment by teijeong** [comment]
Hi lucacampanella@, seems you're using dynamic range quantization for TFLite. From my limited Core ML experience, Core ML provides weight quantization, but when inferencing the weights are dequantized back to FP16 to do calculations. CoreML's linear quantization should result in quite same weights as TFLite's dynamic range quantization, but TFLite quantizes activation and do calculations in int8, so it might be the reason for accuracy loss.

I could provide more debugging tips base on your model, if you could provide one. 

@jianlijianli for more context

**Comment by lucacampanella** [comment]
Hi @teijeong and thanks a lot for the feedback, it's much appreciated.

Here's the code that builds the model:
```
def _build_test_graph(self, with_tf_lite_ready_setup=False):
        # Inputs
        if with_tf_lite_ready_setup:
            # Set the batch size of 1 when using TFLite, or it will result in a much bigger checkpoint that still only works with batch size 1
            batch_size = 1
            inputs = Input(shape=(1), batch_size=batch_size)
        else:
            batch_size = None
            inputs = Input(shape=(None,))

        # Input state
        h_0 = [Input(shape=[self.state_dim], batch_size=batch_size)]

        self.rnn_layer = tensorflow.keras.layers.GRU(self.state_dim, return_state=True)
        rnn_inputs = tf.one_hot(tf.cast(inputs, tf.int32), depth=self.vocab_size)

        embedder = Embedding(self.vocab_size, self.emb_dim, mask_zero=True)
        m = embedder.compute_mask(inputs) # the token 0 is the padding token
        # Apply RNN
        out = self.rnn_layer(rnn_inputs, initial_state=h_0, mask=m)

        rnn_out = out[0]
        h_out = out[1:]

        dense = Dense(self.vocab_size, activation="softmax")
        output = dense(rnn_out)

        return keras.Model([inputs] + h_0, [output] + h_out)
```

So if I understood you correctly, you're saying that TFLite keeps the integer quantization also during inference and thus applies it also to the activations. On the other hand CoreML does inference in float16, which allows the activations to have a broader range of values and yield better accuracy?
What is also interesting to notice is that when I zip the checkpoints of TFLite and CoreML the TFLite checkpoint is smaller. TFLite is 905 kB uncompressed, 505 kB compressed; CoreML (with linear quantization, the default) is 914 kB uncompressed and 840 kB compressed. Do you think this suggests, from a very rough point of view, that there is less information in the TFLite model?

Thanks a lot for the help, any tip or suggestion is welcome!

**Comment by UsharaniPagadala** [comment]
@lucacampanella 

Could you please confirm if the issue is resolved? if yes, please feel free to move this issue to closed status.

**Comment by lucacampanella** [comment]
Hi @UsharaniPagadala ,
I don't really know why this issue was referenced in number #48546 , they seem totally unrelated to me.

The current **issue is not resolved**, I was just waiting for some feedback from @teijeong , who was kind enough to reply to my first message and asked for more code :)

**Comment by lucacampanella** [comment]
Hi @teijeong , did you have a chance to look at the code and at my clarification questions? It would help us a great bunch!
Thanks! :)

**Comment by Venkat6871** [comment]
Hi **@lucacampanella** ,

- TFLite default quantization mode may not be optional for your model. Experiment with different quantization modes such as tf.lite.Optimize.LSTM or tf.lite.Optimize.FLOAT16 to see if they improve accuracy. Since tflite does not directly offer the same quantization modes as coreML, you may need to adjust other parameters or configurations to achieve similar results.
- Ensure that the quantization range used during TFLite conversion is appropriate for your data distribution. You can manually set the quantization range using converter.inference_input/output_type = tf.int8 and converter.inference_input/output_range parameters.

Thank you!
